src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/WorkspaceProjectSnapshotChangeTrigger.cs

```diff
@@ -112,26 +139,77 @@ internal void Workspace_WorkspaceChanged(object sender, WorkspaceChangeEventArgs
             }
         }
 
-        private void EnqueueUpdate(ProjectId projectId)
+        private void InitializeSolution(Solution solution)
+        {
+            Debug.Assert(solution != null);
+
+            foreach (var project in solution.Projects)
+            {
+                if (TryGetProjectSnapshot(project?.FilePath, out var projectSnapshot))
+                {
+                    _workspaceStateGenerator.Enqueue(project, projectSnapshot);
+                }
+            }
+        }
+
+        private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
+        {
+            switch (args.Kind)
+            {
+                case ProjectChangeKind.ProjectAdded:
+                    var associatedWorkspaceProject = _projectManager.Workspace.CurrentSolution.Projects.FirstOrDefault(
+                        project => string.Equals(args.ProjectFilePath, project.FilePath, FilePathComparison.Instance));
+
+                    if (associatedWorkspaceProject != null)
+                    {
+                        _workspaceStateGenerator.Enqueue(associatedWorkspaceProject, args.Newer);
+                    }
+                    break;
+            }
+        }
+
+        private void EnqueueWithDelay(ProjectId projectId)
         {
             // A race is not possible here because we use the main thread to synchronize the updates
             // by capturing the sync context.
             if (!_deferredUpdates.TryGetValue(projectId, out var update) || update.IsCompleted)
             {
-                _deferredUpdates[projectId] = UpdateAfterDelay(projectId);
+                _deferredUpdates[projectId] = EnqueAfterDelay(projectId);
             }
         }
 
-        private async Task UpdateAfterDelay(ProjectId projectId)
+        private async Task EnqueAfterDelay(ProjectId projectId)
         {
-            await Task.Delay(ProjectChangeDelay);
+            await Task.Delay(EnqueueDelay);
 
             var solution = _projectManager.Workspace.CurrentSolution;
             var workspaceProject = solution.GetProject(projectId);
-            if (workspaceProject != null)
+            if (workspaceProject != null && TryGetProjectSnapshot(workspaceProject.FilePath, out var projectSnapshot))
             {
-                _projectManager.WorkspaceProjectChanged(workspaceProject);
+                _workspaceStateGenerator.Enqueue(workspaceProject, projectSnapshot);
             }
         }
+
+        private bool TryGetProjectSnapshot(string projectFilePath, out ProjectSnapshot projectSnapshot)
```

Comment:
Should we add this to the PSM? That seems like a much better place for it?

Reply:
Not sure why I didn't reuse the GetLoadedProject bit from PSM, hmmm, can add this TryGet variant.

Reply:
Ehh, `GetLoadedProject` is sufficient, will just modify this to use `GetLoadedProject` but allow for things like a null project file path.

Reply:
uhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh